File: Tests/Phase 2/Review1/disjont-set/disjoint-set.swift
class UnionFind {
    var root: [Int]
    
    init(size: Int) {
        self.root = Array(0..<size)
    }
    
    func find(_ x: Int) -> Int {
        if root[x] == x {
            return x
        }
        root[x] = find(root[x]) // Path compression
        return root[x]
    }
    
    func union(_ x: Int, _ y: Int) {
        let rootX = find(x)
        let rootY = find(y)
        if rootX != rootY {
            root[rootY] = rootX
        }
    }
    
    func connected(_ x: Int, _ y: Int) -> Bool {
        return find(x) == find(y)
    }
}

// test
let unionFind = UnionFind(size: 10)
unionFind.union(1, 2)
unionFind.union(2, 3)
unionFind.union(4, 5)
unionFind.union(6, 7)
unionFind.union(8, 9)
unionFind.union(1, 9)
print(unionFind.connected(1, 9)) // true
print(unionFind.connected(3, 9)) // true
print(unionFind.connected(1, 4)) // false
print(unionFind.connected(8, 4)) // false
print(unionFind.connected(4, 5)) // true
print(unionFind.connected(5, 6)) // false
print(unionFind.connected(6, 7)) // true

File: Data Structure & Algorithm/Review/Stage 10/Review 1/Part1/disjoint-set/disjoint-set.py
class DisjointSet:
    def __init__(self, size):
        self.root = [i for i in range(size)]

    # The find function here is the same as that in the disjoint set with path compression.
    def find(self, x):
        if self.root[x] == x:
            return x
        return self.find(self.root[x])

    # The union function with union by rank
    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            self.root[root_y] = root_x

    def connected(self, x, y):
        return self.find(x) == self.find(y)


# test
uf = DisjointSet(10)
uf.union(1, 2)
uf.union(2, 3)
uf.union(4, 5)
uf.union(6, 7)
uf.union(8, 9)
uf.union(1, 9)
print(uf.connected(1, 9))
print(uf.connected(3, 9))
print(uf.connected(1, 4))
print(uf.connected(1, 8))
